audit log: Create audit log when a realm is created or imported

hackerkid opened the pull request:
- [x] Add RealmAuditLog for REALM_CREATED

- [x] Backfill REALM_CREATED audit logs

- [x] Add RealmAuditLog for REALM_IMPORTED

- [x] Backfill REALM_IMPORTED log along with from where it was exported(slack, gitter etc)

- [x] Include platform details in zulip export file and store it in REALM_IMPORTED audit log.

timabbott commented on zerver/migrations/0321_realm_creation_log_backfill.py:
```diff
@@ -0,0 +1,38 @@
+    objects_to_create = []
+    for realm in Realm.objects.all():
+        entry = RealmAuditLog(
+            realm=realm, event_type=215, event_time=realm.date_created, backfilled=True
```
For migrations like this, it's best to write `REALM_CREATED`= 215` and then use the variable, so `grep` finds uses of the code; @mateuszmandera can point to the technique.

timabbott commented on zerver/models.py:
```diff
@@ -747,13 +747,22 @@ def presence_disabled(self) -> bool:
+        realm = kwargs["instance"]
+        RealmAuditLog.objects.get_or_create(
+            realm=realm, event_time=realm.date_created, event_type=RealmAuditLog.REALM_CREATED
+        )
```
I'd prefer to just add this logic to `do_create_realm`; we use that whenever creating a realm except in buggy tests anyway, and it's clearer code than using Django signals.

mateuszmandera replied:
Agreed on putting this in do_create_realm - but besides tests, we also do Realm.objects.create in `create_internal_realm` (possibly we should put this RealmAuditLOg creating  code there too?) and in `populate_db`:
```
            # Create our three default realms
            # Could in theory be done via zerver.lib.actions.do_create_realm, but
            # welcome-bot (needed for do_create_realm) hasn't been created yet
            create_internal_realm()
            zulip_realm = Realm.objects.create(
                string_id="zulip",
                name="Zulip Dev",
                emails_restricted_to_domains=False,
                email_address_visibility=Realm.EMAIL_ADDRESS_VISIBILITY_ADMINS,
                description="The Zulip development environment default organization."
                "  It's great for testing!",
                invite_required=False,
                org_type=Realm.CORPORATE,
            )
            RealmDomain.objects.create(realm=zulip_realm, domain="zulip.com")
            if options["test_suite"]:
                mit_realm = Realm.objects.create(
                    string_id="zephyr",
                    name="MIT",
                    emails_restricted_to_domains=True,
                    invite_required=False,
                    org_type=Realm.CORPORATE,
                )
                RealmDomain.objects.create(realm=mit_realm, domain="mit.edu")

                lear_realm = Realm.objects.create(
                    string_id="lear",
                    name="Lear & Co.",
                    emails_restricted_to_domains=False,
                    invite_required=False,
                    org_type=Realm.CORPORATE,
                )
```

This sounds like it should be possible to migrate to use do_create_realm, because the comment "welcome-bot (needed for do_create_realm) hasn't been created yet" doesn't sound accurate anymore (the bot is created together with the internal realm creation). So we should probably do that, unless we want to keep using `.create()` there for some other reasons as well?

hackerkid replied:
Yeah. I think we should use `do_create_realm` there. Let me update the PR.

hackerkid replied:
I think I will create a separate PR for this. There are a lot of unrelated changes for this migration that should be better handled with a different PR.

timabbott commented on zerver/lib/import_realm.py:
```diff
@@ -1213,6 +1213,10 @@ def do_import_realm(import_dir: Path, subdomain: str, processes: int = 1) -> Rea
+
+    RealmAuditLog.objects.create(
+        realm=realm, event_type=RealmAuditLog.REALM_IMPORTED, event_time=timezone_now()
+    )
```
I think this probably belongs before `do_change_plan_type` -- since if that does a `RealmAuditLog` entry, it'll be clear that happened on the new server, not the old.

timabbott replied:
It could be nice to have `extra_data` for this field that records the Zulip server version and feature level?  

Thought perhaps it'd be smarter to try to add that as actual columns in `RealmAuditLog`?  

@mateuszmandera FYI -- do you think that data in RealmAuditLog entries could be helpful in the future when writing migrations?  I guess we already have migration level data from Django's tables, so maybe this would be useless.

(In any case, this idea shouldn't block this PR).

mateuszmandera replied:
@timabbott Hmm I think that could be valuable - e.g. if we find a bug where some code was doing something wrong when making RealmAuditLog-logged changes and we need to migrate the affected objects to have correct data - this could allow a straightfoward way of finding the set of potentially affected objects.

timabbott commented on zerver/models.py:
```diff
@@ -3336,6 +3336,7 @@ class AbstractRealmAuditLog(models.Model):
     REALM_SPONSORSHIP_PENDING_STATUS_CHANGED = 213
     REALM_SUBDOMAIN_CHANGED = 214
     REALM_CREATED = 215
+    REALM_IMPORTED = 216
```
Do we have a theory for how we could identify which realms were imported in the past for a backfilled migration on that front?

I guess the best idea I have on that front is that we could look at the plan_type records for when organizations changed from `SELF_HOSTED`.

Also, it'd be really nice if these records recorded what a realm was imported from (E.g. was it Slack or another Zulip server); so maybe we should adjust the export format to include that detail (if it doesn't already), and then include that in `extra_data` for this PR?

timabbott commented on zerver/views/registration.py:
```diff
@@ -418,6 +419,14 @@ def accounts_register(request: HttpRequest) -> HttpResponse:
+                .last()
+            )
+            realm_creation_audit_log.acting_user = user_profile
+            realm_creation_audit_log.save(update_fields=["acting_user"])
```
This should be in `do_create_user`, if `realm_creation=True`; in additional to being a cleaner abstraction boundary, it also fixes a bug (that `zproject/backends.py` can be used to create a realm with the LDAP auth backend not using this code path).

timabbott commented:
I merged the first commit as 7f3fc3423bb1d657274a6b2cb2df23245a3ac112, after making several changes:
* Added `backfilled=True` flag to the creation events in the import code path
* Moved the block to `do_create_user`
* Fixed the reverse migration not running.
* Tested that the migration runs fast enough to not be disruptive even for Zulip Cloud
* Added `transaction.atomic` around the part where we create the `RealmAuditLog` entry.
Thanks @hackerkid!

For the import event, I did a few tweaks and pushed back here based on testing the queries:
* We've changed the case for the "imported from slack messages" over time, so we need an `iexact` query on the topic.
* We can do a loop, and we should include hipchat even though we've deleted that import tool.

However, that commit still needs work:
* The Gitter query is really expensive/slow and basically hangs on Zulip Cloud. We probably will want to replace it with another tell, e.g. having users with an `@users.noreply.github.com` email address, which isn't possible except via Gitter import.  Not sure.
* The final RealmAuditLog check doesn't work at all -- the `extra_data` string isn't right / correctly quoted.  It's possible we want to try to do @edith007's effort to migrate format to JSONField first so that we can write this reasonably.
* Thinking this through, I guess once this runs, we can manually validate for Zulip Cloud that we believe all the organizations with one of these RealmAuditLog entries were actually imported.

timabbott commented on zerver/migrations/0333_realm_import_audit_log_backfill.py:
```diff
@@ -0,0 +1,101 @@
+            Message.objects.filter(subject__iexact=f"imported from {service}")
+            .values_list("sender__realm", flat=True)
+            .distinct()
+        )
```
A much faster and potentially cleaner tell would be to look at RealmAuditLog and see if their organization description is "Imported from slack" and/or ; but I'm not sure if RealmAuditLog for that property goes back before the import tools, and/or if the import tools always had the same strings for `realm.description`.

hackerkid replied:
> A much faster and potentially cleaner tell would be to look at RealmAuditLog and see if their organization description is "Imported from slack" and/or ; 


You mean we should filter for RealmAuditLogs with `event_type=RealmAuditLog.REALM_PROPERTY_CHANGED` and `extra_data={RealmAuditLog.OLD_VALUE: f"Imported from {service}", "property": "description"}`.

Looks like this audit log was added only in ba92666dbd30e528d6cc23a7bc40c6bc618e252a which merged on on 7 Jul 2020. So I think relying on this is not of much use.

Also, I think we would have to fetch all audit logs of type `RealmAuditLog.REALM_PROPERTY_CHANGED` for a realm, go through them in a loop and check that extra_data. I don't think we can write a query that does that automatically since we are not specifying `RealmAuditLog.NEW_VALUE`. Does @edith007's work on `JSONField` makes queries like this possible?

> if the import tools always had the same strings for realm.description

The commit that properly sets the `realm.description` was added on  Wed Feb 21 14:57:37 2018.
e98dd634151b55a51e7968b8aa8af57487e48cb5. Also it looks like that commit didn't really worked so it looks like 1410a1e46033e32eb8d14eaf979e0f1b16057700 was the commit that actually fixed it on 18 Apr 2018. I think that is also kind of late?

edith007 replied:
yes @hackerkid queries like this will be possible with `JSONField`. With the change, we won't have to fetch all the RealmAuditLogs and loop through each of them to search for a certain extra_data, instead we can give the json or some part of the json or keys in the query itself...

hackerkid commented:
> The Gitter query is really expensive/slow and basically hangs on Zulip Cloud. We probably will want to replace it with another tell, e.g. having users with an @users.noreply.github.com email address, which isn't possible except via Gitter import. Not sure.

We create a stream called "from gitter" when the gitter data is converted. We could check for the existence of this stream. But the problem is the stream might have been renamed. There is a STREAM_NAME_CHANGED audit log which tracks of the stream renaming, but it looks like it was only added 10 months back.

timabbott commented:
OK.  I think the `users.noreply` thing is a really good tell for Gitter; let me verify that.

hackerkid commented:
> The final RealmAuditLog check doesn't work at all -- the extra_data string isn't right / correctly quoted. It's possible we want to try to do @edith007's effort to migrate format to JSONField first so that we can write this reasonably.

I have updated the query to

```python
plan_type_changed_logs = RealmAuditLog.objects.filter(
        event_type=RealmAuditLog.REALM_PLAN_TYPE_CHANGED,
        acting_user=None,
        extra_data=f"{{'old_value': {Realm.SELF_HOSTED}, 'new_value': {Realm.LIMITED}}}",
    )
```

Can your verify the new query works? I have tested it in local dev environment and it's working.

> Hmm, no, it's not great. It's possible we should just do Gitter by hand -- there aren't very many, and I think every fewer into self-hosted servers.


Okay. I have removed the Gitter query. Another query that could have worked was to check whether there are any streams called `from gitter` in the realm. But the issue there is that a lot of them must have been renamed and the audit log that keeps track of the renaming was only added in 54df9290b9660a01cf91b20c7437955c3c5a06bd which was merged on Jul 13 2020

hackerkid commented:
@timabbott This is ready for a review.

timabbott commented on zerver/lib/import_realm.py:
```diff
@@ -871,15 +874,15 @@ def do_import_realm(import_dir: Path, subdomain: str, processes: int = 1) -> Rea
         raise Exception("Missing realm.json file!")
 
-    if not server_initialized():
-        create_internal_realm()
```
I think the `server_initialized` check should remain here, outside the try/except block, since it's unrelated.

timabbott commented on zerver/data_import/gitter.py:
```diff
@@ -338,6 +338,7 @@ def do_convert_data(gitter_data_file: str, output_dir: str, threads: int = 6) ->
     create_converted_data_files([], output_dir, "/uploads/records.json")
     # IO attachments records
     create_converted_data_files(attachment, output_dir, "/attachment.json")
+    create_converted_data_files({"service": "slack"}, output_dir, "/meta.json")
```
What is `meta.json`?  Oh, is this just generating a tiny file with just the `server => slack` field in it?  

I would prefer to store this sort of detail in a field in `realm.json` that isn't imported (and perhaps we should have `manage.py export` set it as well).

hackerkid commented:
Closing since this has became quite stale.
